### ISSUE TITLE
set_intersection doesn't work as expectation - tensorflow 1.6.0

### ISSUE THREAD

**Issue by llv22** [tensorflow/tensorflow#19832]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: NO
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**:
source
- **TensorFlow version (use command below)**:
1.6.0
- **Python version**: 
3.6.5 (anaconda)
- **Bazel version (if compiling from source)**:
0.11
- **GCC/Compiler version (if compiling from source)**:
gcc-5.4.0
- **CUDA/cuDNN version**:
CUDA9.1 and cuDNN 7.1.4
- **GPU model and memory**:
GPU model and 16G memory
- **Exact command to reproduce**:

```python
import collections

a = collections.OrderedDict([
      ((0, 0), 1),
      ((0, 3), 1),
      ((1, 1), 1),
      ((1, 3), 1),
      ((2, 0), 1),
      ((2, 1), 1)
  ])
a = tf.SparseTensor(list(a.keys()), list(a.values()), dense_shape=[3, 4])

b = collections.OrderedDict([
      ((0, 0), 1),
      ((0, 3), 1),
      ((1, 1), 1),
      ((1, 2), 1),
      ((1, 3), 1),
      ((2, 0), 1),
      ((2, 1), 1)
  ])
b = tf.SparseTensor(list(b.keys()), list(b.values()), dense_shape=[3, 4])

with tf.Session() as sess:
    print(sess.run(a))
    print(sess.run(b))
    print(sess.run(tf.contrib.metrics.set_intersection(a, b)))

```

### Describe the problem
For set_intersection, for my own understanding, output should be 

```bash
SparseTensorValue(indices=array([[0, 0],
       [0, 3],
       [1, 1],
       [1, 3],
       [2, 0],
       [2, 1]]), values=array([1, 1, 1, 1, 1, 1], dtype=int32), dense_shape=array([3, 4]))
```

However, I get the result as follow:

```bash
SparseTensorValue(indices=array([[0, 0],
       [1, 0],
       [2, 0]]), values=array([1, 1, 1], dtype=int32), dense_shape=array([3, 1]))
```

I don't understand if I'm not fully understanding for the points....

Best Regards
Orlando

**Comment by VanitarNordic** [comment]
Now, You can use `conda install -c hesi_m keras` to install the latest version of **Keras (2.2.0)** and **Tensorflow(1.8.0)** for you.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @michaelisard: It has been 16 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @michaelisard: It has been 31 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by michaelisard** [comment]
Sorry for the very slow response. @sguada can you help?

**Comment by sguada** [comment]
Not sure maybe @ebrevdo can.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @sguada, @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @sguada, @ebrevdo: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @sguada, @ebrevdo: It has been 50 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @sguada, @ebrevdo: It has been 65 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @sguada, @ebrevdo: It has been 80 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @sguada, @ebrevdo: It has been 95 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @sguada, @ebrevdo: It has been 110 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @sguada, @ebrevdo: It has been 125 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by sushreebarsa** [comment]
@llv22 Is this still an issue ?
We see that you are using old version of tensorflow 1.x which is not actively supported, We recommend that you upgrade to 2.4 or later version.Please have a look at the migration guide for reference to migrate from TensorFlow 1.x to TensorFlow 2.Thank you!

**Comment by llv22** [comment]
@sushreebarsa I will check on tf 2.4 and let you know the result

**Comment by llv22** [comment]
@sushreebarsa I'm afraid that this issue hasn't been settled down.
This is the result given on tf 2.4:
```bash
In [52]: import collections
    ...: 
    ...: a = collections.OrderedDict([
    ...:       ((0, 0), 1),
    ...:       ((0, 3), 1),
    ...:       ((1, 1), 1),
    ...:       ((1, 3), 1),
    ...:       ((2, 0), 1),
    ...:       ((2, 1), 1)
    ...:   ])
    ...: a = tf.SparseTensor(list(a.keys()), list(a.values()), dense_shape=[3, 4])
    ...: 
    ...: b = collections.OrderedDict([
    ...:       ((0, 0), 1),
    ...:       ((0, 3), 1),
    ...:       ((1, 1), 1),
    ...:       ((1, 2), 1),
    ...:       ((1, 3), 1),
    ...:       ((2, 0), 1),
    ...:       ((2, 1), 1)
    ...:   ])
    ...: b = tf.SparseTensor(list(b.keys()), list(b.values()), dense_shape=[3, 4])

In [53]: tf.sparse.to_dense(a).numpy()
Out[53]: 
array([[1, 0, 0, 1],
       [0, 1, 0, 1],
       [1, 1, 0, 0]], dtype=int32)

In [54]: tf.sparse.to_dense(b).numpy()
Out[54]: 
array([[1, 0, 0, 1],
       [0, 1, 1, 1],
       [1, 1, 0, 0]], dtype=int32)

In [55]: tf.sparse.to_dense(tf.sets.intersection(a, b)).numpy()
Out[55]: 
array([[1],
       [1],
       [1]], dtype=int32)

In [68]: tf.print(tf.sets.intersection(a, b))
'SparseTensor(indices=[[0 0]
 [1 0]
 [2 0]], values=[1 1 1], shape=[3 1])'
```

For a, b with the same 3*4 dimension,   their expected intersection based on sparseTensor indices should contain sparseTensor with 6 bits. It seems that the c++ implementation has merged intersection results in the last dimension together and reshaped its size to 1. I suspected it's the right behavior. Could you double-check with it?

In the meanwhile, another trial has also been taken:
```bash
In [59]: a1 = tf.sparse.reshape(a, (3,4,1))

In [60]: b1 = tf.sparse.reshape(b, (3,4,1))

In [61]: tf.print(a1)
'SparseTensor(indices=[[0 0 0]
 [0 3 0]
 [1 1 0]
 [1 3 0]
 [2 0 0]
 [2 1 0]], values=[1 1 1 1 1 1], shape=[3 4 1])'

In [62]: tf.print(b1)
'SparseTensor(indices=[[0 0 0]
 [0 3 0]
 [1 1 0]
 ...
 [1 3 0]
 [2 0 0]
 [2 1 0]], values=[1 1 1 ... 1 1 1], shape=[3 4 1])'

In [67]: tf.print(tf.sets.intersection(a1, b1))
'SparseTensor(indices=[[0 0 0]
 [0 3 0]
 [1 1 0]
 [1 3 0]
 [2 0 0]
 [2 1 0]], values=[1 1 1 1 1 1], shape=[3 4 1])'

```
After transforming both of tensor to 3*4*1, now the result is correct. 
That's really weird.

Could you clarify it further?

Best Regards
Orlando

**Comment by gowthamkpr** [comment]
@llv22 You are right. The c++ implementation has merged intersection results in the last dimension together and reshaped its size to 1.

**Comment by llv22** [comment]
So it will be considered as the expected behavior instead of the abnormal
one?

**Comment by tilakrayal** [comment]
@llv22,
Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

As mentioned it is an intended behaviour for the API. The c++ implementation has merged intersection results in the last dimension together and reshaped its size to 1.
